<h1 class="alert alert-info"><center>Listes Python - Introduction & Opérations élémentaires</center></h1>

##### **_Adapté de C. Poulmaire / AEIF_**

# Exercice 1 : évolution dans le temps du prix d'un article

Dans cet exercice, on considère un article coûtant initialement 50 euros et augmentant de 3,5 euros par an. Nous allons étudier l'évolution du prix au cours du temps.

1. On considère le code Python suivant :

In [1]:
def afficheTousLesPrix(a,n):
    for i in range(1,n+1):
        print(a)
        a = a + 3.5
    return a

afficheTousLesPrix(50,5)

50
53.5
57.0
60.5
64.0


67.5

Compléter ce code de façon à ce que, lors de l'exécution de la cellule, s'affichent tous les prix successifs de l'article sur _"n"_ ans, y compris le prix initial (on doit donc avoir _n + 1_ valeurs !).

On saisit à présent l'instruction `prix = afficheTousLesPrix(50,5)`

Quel est le contenu de la variable `prix` ?

**Elle sera égale à la dernière valeur calculée - 67.5 en l'occurence.**

2. Cette fonction est-elle satisfaisantes pour étudier l'évolution du prix de l'article au cours du temps ?

**Cette fonction ne permet de stocker en mémoire QUE la valeur du prix au bout de N années. Si je veux obtenir la valeur du prix après M (différent de N) années, je dois ré-exécuter la fonction.**

On considère à présent le code Python suivant :

In [1]:
def tousLesPrix(a,n):
    prix = [a]
    for i in range(1,n+1):
        a = a + 3.5
        prix.append(a)
    return prix

tousLesPrix(50,5)

[50, 53.5, 57.0, 60.5, 64.0, 67.5]

3. Exécuter cette cellule. Que renvoie l'instruction `tousLesPrix(50,5)` ?

**Elle renvoie une liste de toutes les valeurs successives qui ont été calculées - on peut donc, une fois cette fonction exécutée, accéder à chacune des valeurs par année, et pas uniquement à la dernière comme dans le cas précédent.**

4. Ecrire dans la cellule ci-dessous une instruction Python donnant le type de la donnée renvoyée par l'instruction `tousLesPrix(50,5)` - reconnaissez-vous ce type ?

In [3]:
type(tousLesPrix(50,5))

<class 'list'>

5. Rendez-vous à présent sur [Python Tutor](https://pythontutor.com/visualize.html#mode=edit); copiez / collez le code ci-dessus et regardez comment se construit cette liste.

- comment se construit la liste `prix` ?
- à quoi sert l'instruction `prix.append(a)` ?
- comment sont repérés les éléments de la liste `prix` ?

**Je vous laisse voir ce que ça donne dans Python Tutor mais l'idée principale ici est que la fonction, comme dans sa première version, calcule année après année la nouvelle valeur du prix, mais qu'au lieu "d'écraser" la valeur précédente dans une variable unique "prix", elle ajoute la nouvelle valeur à la fin d'une liste qu'elle construit pour finalement la retourner; c'est le rôle de l'instruction `prix.append(a)`. Les éléments de la liste sont repérés par leur indice qui va de 0 (le prix initial, `a`) à `n`.**

4. Exécuter la cellule ci-dessous.

In [3]:
listePrix = tousLesPrix(50,5)
len(listePrix)

6

- Exécuter la cellule ci-dessus. Que renvoie la fonction `len` ?

**Elle renvoie la longueur de la liste (le nombre de valeurs qu'elle comporte) - soit, ici, 6.**

5. Ecrire dans les deux cellules ci-dessous une instruction renvoyant le premier élément et une instruction renvoyant le dernier élément de la liste `listePrix`.

In [4]:
# Ecrire ici une instruction Python renvoyant le premier élément de listePrix (pour rappel du cours, le format est Liste[indice])
listePrix[0]

50

In [6]:
# Ecrire ici une instruction Python renvoyant le dernier élément de listePrix
listePrix[len(listePrix) - 1]
# ATTENTION ICI: les indices démarrent à 0, donc le dernier est bien len(listePrix) - 1!
# C'est ce qui explique la réponse à la question suivante.

67.5

- Exécuter la cellule ci-dessous. Qu'obtient-on ? Pourquoi ?

In [7]:
listePrix[6]

Traceback (most recent call last):
  File "<input>", line 1, in <module>
IndexError: list index out of range


Error: 

**On obtient une erreur d'index puisqu'on demande un élément qui est au-delà de la fin de la liste.**

# Exercice 2 : définition d'une liste par extension

1. Compléter le code suivant de telle sorte que la liste `L` définie à la première ligne devienne `[1,5,8,-3]`.

In [8]:
L = [1,5,8]
L.append(-3)

2. Compléter le code suivant de telle sorte que la liste `L` définie à la première ligne devienne `[2,5,4,2]`.

In [9]:
L = [2,5]
L.append(4)
L.append(2)

3. Compléter le code suivant de telle sorte que la liste `L` définie à la première ligne devienne `[0,1,4,9,16,25]`.

In [10]:
L = []
for i in range(6):
    L.append(i ** 2) # Chaque valeur de la liste est le carré de son indice (0 au carré, 1 au carré, 2 au carré, etc...)

# Exercice 3 : nombre de licenciés d'un club de triathlon

Un club de triathlon compte $1500$ licenciés. Chaque année, il perd $5\%$ de ses licenciés et enregistre $400$ nouveaux inscrits.

Compléter la fonction suivante pour qu'elle renvoie une liste contenant l'évolution du nombre de licenciés sur un nombre `nbAnnees` d'années.

In [16]:
def nbLicenciesTriathlon(nbAnnees):
    listeNbLicencies = [1500]
    for i in range(nbAnnees):
        listeNbLicencies.append(int(listeNbLicencies[i] * 0.95 + 400)) # on calcule un nombre entier de licenciés !
    return listeNbLicencies

# Vous noterez ici que chaque itération de la boucle calcule la prochaine valeur
# sur la base de la valeur qu'elle vient de calculer (par le biais de listeNbLicencies[i])

Les quatre cellules suivantes permettent de tester votre fonction. Elles doivent renvoyer `True` lors de leur exécution.

In [12]:
nbLicenciesTriathlon(0) == [1500]

True

In [13]:
nbLicenciesTriathlon(1) == [1500, 1825]

True

In [14]:
nbLicenciesTriathlon(4) == [1500, 1825, 2133, 2426, 2704]

True

In [15]:
nbLicenciesTriathlon(8) == [1500, 1825, 2133, 2426, 2704, 2968, 3219, 3458, 3685]

True

# Exercice 4 : échange d'éléments dans une liste

La fonction `echange` ci-dessous prend en paramètre une liste non vide `L` et deux entiers `i` et `j`, et renvoie la liste `L` dans laquelle les éléments d'indices `i` et `j` ont été échangés.

Avant d'effectuer l'échange des deux éléments, la fonction effectue trois tests d'assertion :
 - le premier vérifie que `L` est bien une liste non vide ;
 - le deuxième vérifie que l'entier `i` est bien un nombre compatible avec un indice de `L` ;
 - le troisième vérifie que l'entier `j` est bien un nombre compatible avec un indice de `L`.

Compléter cette fonction `echange`.

In [19]:
def echange(L,i,j):
    temp = L[i]
    L[i] = L[j]
    L[j] = temp
    return L

Les trois cellules suivantes permettent de tester votre fonction. Elles doivent renvoyer `True` lors de leur exécution.

In [20]:
echange([8,9],0,1) == [9,8]

True

In [21]:
echange([1,2,3,5,4,6,7],3,4) == [1,2,3,4,5,6,7]

True

In [22]:
echange(['A','C','B','D'],1,2) == ['A','B','C','D']

True

# Exercice 5 : opérateurs de concaténation et de duplication

Déterminer ce que renvoie chacune des instructions suivantes puis vérifier en exécutant les cellules :

In [23]:
[1,2,3] + [4,5]

[1, 2, 3, 4, 5]

In [24]:
[-1,2,3] + [0]

[-1, 2, 3, 0]

In [25]:
[5,4,3] + []

[5, 4, 3]

In [26]:
[] + [-1,2] 

[-1, 2]

In [27]:
[5,4,3] + 5

Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: can only concatenate list (not "int") to list


Error: 

In [28]:
[4,2,6] + [-1,3] + [-5,6,7]

[4, 2, 6, -1, 3, -5, 6, 7]

In [29]:
[5]*2

[5, 5]

In [30]:
6*[0]

[0, 0, 0, 0, 0, 0]

In [31]:
[0,8,9] + [6]*3

[0, 8, 9, 6, 6, 6]

# Exercice 6 : insertion d'élément dans une liste

1. La fonction `insertionDebutListe` ci-dessous prend en paramètre une liste `L` et une donnée `element`, et renvoie une liste dans laquelle `element` a été inséré au début de `L` à l'aide de l'opérateur de concaténation.

Compléter cette fonction.

In [36]:
def insertionDebutListe(L,element):
    return [element] + L

Les trois cellules ci-dessous sont un jeu de tests pour tester votre fonction; elles doivent toutes renvoyer True.

In [37]:
insertionDebutListe([],5) == [5]

True

In [38]:
insertionDebutListe([4],0) == [0,4]

True

In [39]:
insertionDebutListe([2,1,5],-3) == [-3,2,1,5]

True

2. En s'aidant de la fonction `insertionDebutListe` écrite ci-dessus, écrire une fonction `insertionFinListe` qui permet d'insérer un élément à la fin d'une liste.

In [43]:
# Ecrire ici le code de la fonction insertionFinListe
def insertionFinListe(L,element):
    return L + [element]
    # Note: un simple "append" aurait aussi fonctionné.... "L.append(element)"

Les trois cellules ci-dessous sont un jeu de tests pour tester votre fonction; elles doivent toutes renvoyer True.

In [44]:
insertionFinListe([],5) == [5]

True

In [45]:
insertionFinListe([4],0) == [4,0]

True

In [46]:
insertionFinListe([2,1,5],-3) == [2,1,5,-3]

True

3. Reprenez votre fonction `insertionDebutListe` ci-dessus et complétez-la pour qu'elle n'insère l'élément demandé _que_ s'il n'est pas déjà présent dans la liste (ainsi, un appel à `insertionDebutListe([1, 2],0)` renverra `[0, 1, 2]` tandis qu'un appel à `insertionDebutListe([1, 2],1)` renverra `[1, 2]` - donc la liste inchangée).

In [1]:
# Ecrire ici le code de la fonction insertionDebutListe
def insertionDebutListe(L,element):
    if element in L: # On utilise l'opérateur "in" qui permet de faire un test d'appartenance.
        return L
    else:
        return [element] + L

Les trois cellules ci-dessous sont un jeu de tests pour tester votre fonction; elles doivent toutes renvoyer True.

In [2]:
insertionDebutListe([1, 2],9) == [9, 1, 2]

True

In [3]:
insertionDebutListe([1, 2],1) == [1, 2]

True

In [4]:
insertionDebutListe([1, 2],2) == [1, 2]

True